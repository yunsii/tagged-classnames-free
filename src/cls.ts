import clsx from 'clsx'

function normalizeStrings(strings: string[]) {
  let result: string[] = []

  strings.forEach((stringItem) => {
    const segments = stringItem
      .split('\n')
      .map((item) => item.trim())
      .filter((item) => !item.startsWith('//'))

    segments.forEach((item) => {
      result = result.concat(
        item
          .split(/[\t ]/)
          .map((item) => item.trim())
          .filter(Boolean),
      )
    })
  })
  return result.join(' ')
}

export function handleCls(
  strings: TemplateStringsArray,
  ...expressions: any[]
) {
  const classNamesList = strings.reduce((prev, current, currentIndex) => {
    const expression = expressions[currentIndex] || ''
    prev.push(current, clsx(expression))
    return prev
  }, [] as string[])
  return normalizeStrings(classNamesList)
}

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates
export function cls(strings: TemplateStringsArray, ...expressions: any[]) {
  return handleCls(strings, ...expressions)
}
