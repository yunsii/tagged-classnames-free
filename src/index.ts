export { cls } from './cls'
export { tw } from './tw'
