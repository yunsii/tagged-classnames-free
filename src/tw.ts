import { twMerge } from 'tailwind-merge'

import { handleCls } from './cls'

export function tw(strings: TemplateStringsArray, ...expressions: any[]) {
  const clsResult = handleCls(strings, ...expressions)
  return twMerge(clsResult)
}
