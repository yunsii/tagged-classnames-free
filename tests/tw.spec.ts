import { expect, it } from 'vitest'

import { tw } from '../src'

it('basic', () => {
  const result = tw`text-sm`
  expect(result).toMatchSnapshot()
})

it('basic with expressions', () => {
  const result = tw`text-sm ${'font-bold'}   ${'bg-sky-500'}`
  expect(result).toMatchSnapshot()
})

it('multiple lines', () => {
  const result = tw`
    text-sm  bg-sky-500
    font-bold
    text-lg
  `
  expect(result).toMatchSnapshot()
})

it('multiple lines with expressions', () => {
  const result = tw`
    text-sm              bg-sky-400
    font-bold ${'font-bold'}   ${'bg-sky-500'}
  `
  expect(result).toMatchSnapshot()
})

it('multiple lines with comments', () => {
  const result = tw`
    // hello
    text-sm  bg-sky-500
    font-bold
    // world
    text-lg
  `
  expect(result).toMatchSnapshot()
})
