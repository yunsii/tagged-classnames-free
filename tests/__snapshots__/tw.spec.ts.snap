// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `"text-sm"`;

exports[`basic with expressions 1`] = `"text-sm font-bold bg-sky-500"`;

exports[`multiple lines 1`] = `"bg-sky-500 font-bold text-lg"`;

exports[`multiple lines with comments 1`] = `"bg-sky-500 font-bold text-lg"`;

exports[`multiple lines with expressions 1`] = `"text-sm font-bold bg-sky-500"`;
